receiver/hostmetricsreceiver/internal/scraper/cpuscraper/cpu_scraper.go

```diff
@@ -0,0 +1,133 @@
+	"github.com/open-telemetry/opentelemetry-collector/consumer/pdatautil"
+	"github.com/open-telemetry/opentelemetry-collector/internal/data"
+)
+
+// Scraper for CPU Metrics
+type Scraper struct {
+	config    *Config
+	consumer  consumer.MetricsConsumer
+	startTime time.Time
+}
+
+// NewCPUScraper creates a set of CPU related metrics
+func NewCPUScraper(ctx context.Context, cfg *Config, consumer consumer.MetricsConsumer) (*Scraper, error) {
+	return &Scraper{config: cfg, consumer: consumer}, nil
+}
+
+// Start
+func (c *Scraper) Start(ctx context.Context) error {
+	go func() {
+		c.startTime = time.Now()
+
+		ticker := time.NewTicker(c.config.CollectionInterval())
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ticker.C:
+				c.scrapeMetrics(ctx)
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+// Close
+func (c *Scraper) Close(ctx context.Context) error {
+	return nil
+}
+
+func (c *Scraper) scrapeMetrics(ctx context.Context) {
+	ctx, span := trace.StartSpan(ctx, "cpuscraper.scrapeMetrics")
+	defer span.End()
+
+	metricData := data.NewMetricData()
+	metrics := pdatautil.InitializeMetricSlice(metricData)
+
+	err := c.scrapeAndAppendMetrics(metrics)
+	if err != nil {
+		span.SetStatus(trace.Status{Code: trace.StatusCodeDataLoss, Message: fmt.Sprintf("Error(s) when scraping cpu metrics: %v", err)})
+		return
+	}
+
+	if metrics.Len() > 0 {
+		err := c.consumer.ConsumeMetrics(ctx, pdatautil.MetricsFromInternalMetrics(metricData))
+		if err != nil {
+			span.SetStatus(trace.Status{Code: trace.StatusCodeDataLoss, Message: fmt.Sprintf("Unable to process metrics: %v", err)})
+			return
+		}
+	}
+}
+
+func (c *Scraper) scrapeAndAppendMetrics(metrics pdata.MetricSlice) error {
+	cpuTimes, err := cpu.Times(c.config.ReportPerCPU)
+	if err != nil {
+		return err
+	}
+
+	metric := pdatautil.AddMetric(metrics)
+	initializeMetricCPUSecondsFrom(cpuTimes, c.startTime, metric)
+	return nil
+}
+
+const gopsCPUTotal string = "cpu-total"
+
+func initializeMetricCPUSecondsFrom(cpuTimes []cpu.TimesStat, startTime time.Time, metric pdata.Metric) {
+	InitializeMetricCPUSecondsDescriptor(metric.MetricDescriptor())
+
+	idps := metric.Int64DataPoints()
+	idps.Resize(5 * len(cpuTimes))
+	for i, cpuTime := range cpuTimes {
+		labels := map[string]string{}
+		if cpuTime.CPU != gopsCPUTotal {
+			labels[CPULabel] = cpuTime.CPU
+		}
+
+		for j, stateTime := range []struct {
+			label   string
+			timeVal float64
+		}{
+			{UserStateLabelValue, cpuTime.User},
+			{SystemStateLabelValue, cpuTime.System},
+			{IdleStateLabelValue, cpuTime.Idle},
+			{InterruptStateLabelValue, cpuTime.Irq},
+			{IowaitStateLabelValue, cpuTime.Iowait},
+		} {
+			idp := idps.At(i*5 + j)
+			labels[StateLabel] = stateTime.label
+			idp.LabelsMap().InitFromMap(labels)
+			idp.SetStartTime(pdata.TimestampUnixNano(uint64(startTime.UnixNano())))
+			idp.SetTimestamp(pdata.TimestampUnixNano(uint64(time.Now().UnixNano())))
+			idp.SetValue(int64(stateTime.timeVal))
+		}
+	}
```

Comment:
I have a super hard time reading this code. Can you please simplify?

Reply:
Yea sorry this was a bit dirty. Should be much easier to read now,